pallets/inflation/src/lib.rs

```diff
@@ -0,0 +1,604 @@
+    impl<T: Config> Hooks<BlockNumber> for Pallet<T> {
+        fn on_initialize(now: BlockNumber) -> Weight {
+            let recalculation_weight =
+                if Self::is_recalculation_in_next_block(now, &ActiveInflationConfig::<T>::get()) {
+                    T::WeightInfo::hook_with_recalculation()
+                } else {
+                    T::WeightInfo::hook_without_recalculation()
+                };
+
+            // Benchmarks won't acount for whitelisted storage access so this needs to be added manually.
+            //
+            // SafetyInflationTracker - 1 DB read & write
+            // ActiveInflationConfig - 1 DB read
+            let whitelisted_weight =
+                <T as frame_system::Config>::DbWeight::get().reads_writes(2, 1);
+
+            recalculation_weight
+                .saturating_add(T::WeightInfo::on_timestamp_set())
+                .saturating_add(whitelisted_weight)
+        }
+
+        fn on_finalize(now: BlockNumber) {
+            // Recalculation is done at the block right before the re-calculation is supposed to happen.
+            // This is to ensure all the rewards are paid out according to the new inflation configuration from next block.
+            //
+            // If this was done in `on_initialize`, collator & treasury would receive incorrect rewards for that one block.
+            //
+            // This should be done as late as possible, to ensure all operations that modify issuance are done.
+            if Self::is_recalculation_in_next_block(now, &ActiveInflationConfig::<T>::get()) {
+                let (max_emission, config) = Self::recalculate_inflation(now);
+                ActiveInflationConfig::<T>::put(config.clone());
+
+                SafetyInflationTracker::<T>::mutate(|tracker| {
+                    tracker.cap.saturating_accrue(max_emission);
+                });
+
+                Self::deposit_event(Event::<T>::NewInflationConfiguration { config });
+            }
+        }
+
+        fn integrity_test() {
+            assert!(T::CycleConfiguration::periods_per_cycle() > 0);
+            assert!(T::CycleConfiguration::eras_per_voting_subperiod() > 0);
+            assert!(T::CycleConfiguration::eras_per_build_and_earn_subperiod() > 0);
+            assert!(T::CycleConfiguration::blocks_per_era() > 0);
+        }
+    }
+
+    impl<Moment, T: Config> OnTimestampSet<Moment> for Pallet<T> {
+        fn on_timestamp_set(_moment: Moment) {
+            let amount = Self::payout_block_rewards();
+
+            // Update the tracker, but no check whether an `issued` has exceeded `cap`.
+            // This can modified if needed, but these amounts are supposed to be small &
+            // collators need to be paid for producing the block.
+            // TODO: potential discussion topic for the review!
+            SafetyInflationTracker::<T>::mutate(|tracker| {
+                tracker.issued.saturating_accrue(amount);
+            });
```

Comment:
I know this will be ensured by tests that this'll not happen, but what if it happens then even for the next blocks we might be issuing more than cap, though unlikely as payout needs to be more than emission, right?

Reply:
You're right this must be ensured by tests.
The config is static enough, and calculation is quite predictable so I don't believe we'll ever be seeing a scenario when this overflows.

In a scenario when this happens, collator and treasury rewards would keep being issued.
But staker & dApp rewards would no longer be claimable. This would be a big deal, an issue would be raised and we'd have to deal with it.

At least, that was my idea when writing this 😄